compiler/include/symbol.h

```diff
@@ -361,6 +379,45 @@ class ArgSymbol : public LcnSymbol {
 *                                                                   *
 ********************************* | ********************************/
 
+class ShadowVarSymbol : public VarSymbol {
+public:
+  ShadowVarSymbol(ForallIntentTag iIntent, const char* iName, Expr* iSpec = NULL);
+
+  virtual void    verify();
+  virtual void    accept(AstVisitor* visitor);
+  DECLARE_SYMBOL_COPY(ShadowVarSymbol);
+
+  virtual void    replaceChild(BaseAST* oldAst, BaseAST* newAst);
+  virtual bool    isConstant()                              const;
+  virtual bool    isConstValWillNotChange()                 const;
+
+  const char* intentDescrString() const;
+  // The corresponding outer var or NULL if not applicable.
+  SymExpr* outerVarSE() const { return (SymExpr*)outerVarRep; }
+  bool     isReduce()   const { return intent == TFI_REDUCE;  }
+  Expr*    spec()       const;
+  void     removeSupportingReferences();
```

Comment:
Is this method basically implementing some part of lowering? Or ... ? At least we need a comment about it.

Reply:
You are right. I will add this:
  // Remove no-longer-needed references to outside symbols when lowering.